library/ccm.c

```diff
@@ -422,27 +423,32 @@ int mbedtls_ccm_self_test( int verbose )
         if( verbose != 0 )
             mbedtls_printf( "  CCM-AES #%u: ", (unsigned int) i + 1 );
 
+        memset( plaintext, 0, 24);
+        memset( cipertext, 0, 32);
+        memcpy( plaintext, msg, msg_len[i]);
+
         ret = mbedtls_ccm_encrypt_and_tag( &ctx, msg_len[i],
-                                   iv, iv_len[i], ad, add_len[i],
-                                   msg, out,
-                                   out + msg_len[i], tag_len[i] );
+                                           iv, iv_len[i], ad, add_len[i],
+                                           plaintext, cipertext,
+                                           cipertext + msg_len[i], tag_len[i] );
 
         if( ret != 0 ||
-            memcmp( out, res[i], msg_len[i] + tag_len[i] ) != 0 )
+            memcmp( cipertext, res[i], msg_len[i] + tag_len[i] ) != 0 )
         {
             if( verbose != 0 )
                 mbedtls_printf( "failed\n" );
 
             return( 1 );
         }
+        memset( plaintext, 0, 24 );
 
         ret = mbedtls_ccm_auth_decrypt( &ctx, msg_len[i],
-                                iv, iv_len[i], ad, add_len[i],
-                                res[i], out,
-                                res[i] + msg_len[i], tag_len[i] );
+                                        iv, iv_len[i], ad, add_len[i],
+                                        cipertext, plaintext,
+                                        res[i] + msg_len[i], tag_len[i] );
```

Comment:
Shouldn't this point to the tag written during the encryption, just for completeness sake?

Reply:
This was my initial thought. This is the tag that is compared with the generated tag. Since the tag generated in encryption is compared with this value, after the encryption, it doesn't really matter. 
In real use case, the tag that is compared is from the result in the encryption, so I tend to agree with you.